### PR TITLE
Improve function calling (close #279)

### LINKED ISSUE #279: [Feature Request] Improve function calling implementation
### Required prerequisites

- [X] I have searched the [Issue Tracker](https://github.com/camel-ai/camel/issues) and [Discussions](https://github.com/camel-ai/camel/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [X] Consider asking first in a [Discussion](https://github.com/camel-ai/camel/discussions/new).

### Motivation

The current function calling is implemented by parsing the docstring to JSON schema which causes inconsistency in docstring styles. A promising way to improve it is using `pydantic` like `instructor` https://github.com/jxnl/instructor.

### Solution

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by forever-ly**
## Description

As metioned in [[#279](https://github.com/camel-ai/camel/issues/279)](https://github.com/camel-ai/camel/issues/279), the parsing of functions relies on specific annotation styles(Google-style). To support a wider range of docstring styles, , I have reimplemented the function `camel.utils.functions.parse_doc` by introducing `docstring_parser` package(https://pypi.org/project/docstring-parser/) after disscusing with [[@lightaime](https://github.com/lightaime)](https://github.com/lightaime). Currently version supports ReST, Google, Numpydoc-style and Epydoc docstrings.

## Motivation and Context

close #279 

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by Billy1900 — APPROVED** [review]

File: camel/utils/functions.py [hunk 1]
```diff
@@ -151,55 +142,56 @@ def parse_doc(func: Callable) -> Dict[str, Any]:
         Dict[str, Any]: A dictionary with the function's name,
```
Comment (Wendong-Fan) [inline camel/utils/functions.py:142]: Docstring between `Args:` and `Returns:` missing one empty line

File: pyproject.toml [hunk 2]
```diff
@@ -48,6 +48,7 @@ soundfile = { version = "^0", optional = true }
 sentencepiece = { version = "^0", optional = true }
 opencv-python = { version = "^4", optional = true }
 wikipedia = { version = "^1", optional = true }
+docstring-parser = "^0.15"
```
Comment (Wendong-Fan) [inline pyproject.toml:51]: move this upper if it's not an optional package

**Review by Wendong-Fan — APPROVED** [review]
@forever-ly Thanks for the great work! only 3 small format issues to fix, one of the format issues is in the test code which caused the pre-commit fail

**Review by FUYICC — APPROVED** [review]
It looks great!

**Comment by Wendong-Fan** [comment]
hey @Billy1900 @FUYICC ,

Just noticed this is the old PR, the latest one is https://github.com/camel-ai/camel/pull/393
Please review the latest one if you have time, I will close this PR directly and sorry for the confusion

**Comment by lightaime** [comment]
Closed since this is a duplicated with https://github.com/camel-ai/camel/pull/393.

**Pull request closed without merging**